[Trivia] V3 Rewrite

Tobotimus opened the pull request:
This PR supersedes #789 and as such many of the following feature suggestions came from there.

### Goals
- [x] Get cog working in V3
- [x] Rewrite `TriviaSession`
- [x] Consolidate how the trivia lists are stored and read (#917)
- [x] Add a payout option for winning
- [x] Use YAML for trivia lists

### Possible enhancements
- [x] Multiple categories per session
- [x] Implement some sort of leaderboard system
- [ ] Use of embeds
- [ ] MCQ trivia

Tobotimus committed:
Rewritten basic settings

Tobotimus committed:
Ported to V3

Tobotimus committed:
Will saves the dicts

Tobotimus committed:
TriviaSession uses Config

Tobotimus committed:
wait_for is the future

Tobotimus committed:
Out with recursion and redundancy

Tobotimus committed:
Remove typing before first question

Tobotimus commented:
Alright, `TriviaSession` has just undergone some major changes, and I'm interested to see what you guys think. The back-end of this cog is now fairly complete, so please... try to break it :smile:

_Edit: I probably should have mentioned that although most of the back-end is complete, we are still waiting to see what we are doing with trivia lists._

Tobotimus committed:
Added payout multiplier setting

Tobotimus committed:
Payout setting uses appropriate check

Tobotimus committed:
Implement actual paying of winner

Twentysix26 commented:
I think you can safely remove the trivia lists from your PR in the meantime. It's unlikely they'll be kept in this branch and they can eventually be added with a different pull request if needed

Tobotimus committed:
Removed trivia lists from repository

Tobotimus commented:
That's fair enough, they were only there for testing purposes, so no real reason for them to be in the repository, considering they would only be there temporarily.

But on that note, if anyone is testing this PR, trivia lists must be located in the `cogs/trivia/lists` folder in order to be found (until further notice).

Tobotimus committed:
Added payout confirmation message

Also rearranged the order of methods in TriviaSession to better represent the sequential order of running trivia

Tobotimus committed:
Contestant count fixed

Also fixed error when lists folder is missing

Tobotimus committed:
Support for multiple categories per session

Tobotimus committed:
Extra error handling

Tobotimus committed:
Update from rebase

Tobotimus committed:
Use of task cancel

Tobotimus committed:
Nicer docstrings

Tobotimus committed:
Oops

Tobotimus committed:
Better imports

Tobotimus committed:
YAML lists

Tobotimus committed:
Using the trivia list repo (YAY)

calebj commented on redbot/cogs/trivia/session.py:
```diff
@@ -0,0 +1,229 @@
+        except asyncio.TimeoutError:
+            if abs(self._last_response - int(time.perf_counter())) >= timeout:
+                await self.ctx.send("Guys...? Well, I guess I'll stop then.")
+                await self.stop()
```
`self.stop()` isn't an async function.

calebj commented on redbot/cogs/trivia/session.py:
```diff
@@ -0,0 +1,229 @@
+
+        """
+        try:
+            message = await self.ctx.bot.wait_for(
```
I don't see anything here that would cause the issue you described. Try to debug and isolate it.

Tobotimus committed:
Cast all answers to string since YAML is dumb

Tobotimus committed:
session.stop() is not an async function

Tobotimus committed:
Remove redundant attribute from session

Tobotimus committed:
Sessions manage their own tasks

Tobotimus committed:
Add `redbot/trivia` to .gitignore

Tobotimus committed:
Add PyYAML to requirements.txt

Tobotimus commented:
I think this PR is ready for proper testing. Any extra features should probably be added in separate PRs.

To test right now, I recommend doing the following in a clean virtual environment:
```
pip install --process-dependency-links git+https://github.com/Tobotimus/Red-DiscordBot.git@V3/trivia_rewrite
```
Then run redbot as usual with `redbot <instance_name>`

tekulvw reviewed (changes requested):
Can you add `Red-Trivia` to the requirements for the bot? I'll get it uploaded to pypi soon.

Tobotimus committed:
Parse answers properly before checking

Tobotimus committed:
Add Red-Trivia to reqs

Tobotimus committed:
Better task management and some optimisations

Sessions now just use a flat dict for settings instead of config. This means the settings can't be changed after the session is started, but it removes the need for config to be read for every question.

Tobotimus committed:
Allow lists to override settings

Tobotimus committed:
Fix config logic

Tobotimus committed:
Iteration fixes, config override fixes, task management fixes

Tobotimus committed:
Don't hide dict abuse

Tobotimus committed:
Stats tracking

Tobotimus committed:
Merge V3/develop into V3/trivia_rewrite

Tobotimus committed:
Leaderboard implemented

Tobotimus committed:
Merge branch 'V3/develop' into V3/trivia_rewrite

Tobotimus committed:
[leaderboard] ignore self and reverse list

Tobotimus committed:
[leaderboard] Handle empty score counter

tekulvw committed:
Merge branch 'V3/develop' into V3/trivia_rewrite

Tobotimus committed:
Only count win when max score is reached

palmtree5 committed:
Merge branch 'V3/develop' into V3/trivia_rewrite
